Remove support for overlay/overlay2 without d_type

thaJeztah opened the pull request:
This pull request contains three changes related to `d_type` support on overlay/overlay2:

### Perform fsmagic detection on driver's home-dir if it exists

The fsmagic check was always performed on "data-root" (`/var/lib/docker`), not on the storage-driver's home directory (e.g. `/var/lib/docker/<somedriver>`).

This caused detection to be done on the wrong filesystem in situations where `/var/lib/docker/<somedriver>` was a mount, and a different filesystem than `/var/lib/docker` itself.

This patch checks if the storage-driver's home directory exists, and only falls back to `/var/lib/docker` if it doesn't exist.


## Remove support for overlay/overlay2 without d_type

Support for running overlay/overlay2 on a backing filesystem without d_type support (most likely: xfs, as ext4 supports this by default), was deprecated for some time.

Running without d_type support is problematic, and can lead to difficult to debug issues ("invalid argument" errors, or unable to remove files from the container's filesystem); see https://github.com/moby/moby/issues/10294 and https://github.com/moby/moby/issues/29873 for examples.

This patch turns the warning that was previously printed into an "unsupported" error, so that the overlay/overlay2 drivers are not automatically selected when detecting supported storage drivers.


## Allow existing setups to continue using d_type

Even though it's highly discouraged, there are existing installs that are running overlay/overlay2 on filesystems without d_type support.

This patch allows the daemon to start in such cases, instead of refusing to start without an option to override.

For fresh installs, backing filesystems without d_type support will still cause the overlay/overlay2 drivers to be marked as "unsupported", and skipped during the automatic selection.

This feature is only to keep backward compatibility, but should be removed at some point.


**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

```Markdown
* Disable support for overlay and overlay2 storage drivers on backing filesystems without `d_type` support [moby/moby#35514](https://github.com/moby/moby/pull/35514)
```

thaJeztah committed:
Remove support for overlay/overlay2 without d_type

Support for running overlay/overlay2 on a backing filesystem
without d_type support (most likely: xfs, as ext4 supports
this by default), was deprecated for some time.

Running without d_type support is problematic, and can
lead to difficult to debug issues ("invalid argument" errors,
or unable to remove files from the container's filesystem).

This patch turns the warning that was previously printed
into an "unsupported" error, so that the overlay/overlay2
drivers are not automatically selected when detecting supported
storage drivers.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Power: https://jenkins.dockerproject.org/job/Docker-PRs-powerpc/7020/console

```
20:24:39 --- FAIL: TestOverlaySetup (0.00s)
20:24:39 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-427789185/overlay: no such file or directory
20:24:39 	graphtest_unix.go:48: open /tmp/docker-graphtest-427789185/overlay: no such file or directory
20:24:39 --- FAIL: TestOverlayCreateEmpty (0.00s)
20:24:39 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-718595820/overlay: no such file or directory
20:24:39 	graphtest_unix.go:48: open /tmp/docker-graphtest-718595820/overlay: no such file or directory
20:24:39 --- FAIL: TestOverlayCreateBase (0.00s)
20:24:39 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-933558107/overlay: no such file or directory
20:24:39 	graphtest_unix.go:48: open /tmp/docker-graphtest-933558107/overlay: no such file or directory
20:24:39 --- FAIL: TestOverlayCreateSnap (0.00s)
20:24:39 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-064331006/overlay: no such file or directory
20:24:39 	graphtest_unix.go:48: open /tmp/docker-graphtest-064331006/overlay: no such file or directory
20:24:39 --- FAIL: TestOverlay50LayerRead (0.00s)
20:24:39 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-500077637/overlay: no such file or directory
20:24:39 	graphtest_unix.go:48: open /tmp/docker-graphtest-500077637/overlay: no such file or directory
20:24:39 FAIL
20:24:39 coverage: 10.3% of statements
20:24:39 FAIL	github.com/docker/docker/daemon/graphdriver/overlay	0.006s
20:24:39 --- FAIL: TestOverlaySetup (0.00s)
20:24:39 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-288586816/overlay2: no such file or directory
20:24:39 	graphtest_unix.go:48: open /tmp/docker-graphtest-288586816/overlay2: no such file or directory
20:24:39 --- FAIL: TestOverlayCreateEmpty (0.00s)
20:24:39 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-165054623/overlay2: no such file or directory
20:24:39 	graphtest_unix.go:48: open /tmp/docker-graphtest-165054623/overlay2: no such file or directory
20:24:39 --- FAIL: TestOverlayCreateBase (0.00s)
20:24:39 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-387230066/overlay2: no such file or directory
20:24:39 	graphtest_unix.go:48: open /tmp/docker-graphtest-387230066/overlay2: no such file or directory
20:24:39 --- FAIL: TestOverlayCreateSnap (0.00s)
20:24:39 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-181501481/overlay2: no such file or directory
20:24:39 	graphtest_unix.go:48: open /tmp/docker-graphtest-181501481/overlay2: no such file or directory
20:24:39 --- FAIL: TestOverlay128LayerRead (0.00s)
20:24:39 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-471997812/overlay2: no such file or directory
20:24:39 	graphtest_unix.go:48: open /tmp/docker-graphtest-471997812/overlay2: no such file or directory
20:24:39 --- FAIL: TestOverlayDiffApply10Files (0.02s)
20:24:39 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-873032045/overlay2: no such file or directory
20:24:39 	graphtest_unix.go:48: open /tmp/docker-graphtest-873032045/overlay2: no such file or directory
20:24:39 --- FAIL: TestOverlayChanges (0.00s)
20:24:39 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-536687143/overlay2: no such file or directory
20:24:39 	graphtest_unix.go:48: open /tmp/docker-graphtest-536687143/overlay2: no such file or directory
20:24:39 FAIL
20:24:39 coverage: 13.5% of statements
```

Z: https://jenkins.dockerproject.org/job/Docker-PRs-s390x/6824/console

```
20:25:53 --- FAIL: TestOverlaySetup (0.00s)
20:25:53 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-308327245/overlay: no such file or directory
20:25:53 	graphtest_unix.go:48: open /tmp/docker-graphtest-308327245/overlay: no such file or directory
20:25:53 --- FAIL: TestOverlayCreateEmpty (0.00s)
20:25:53 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-728787528/overlay: no such file or directory
20:25:53 	graphtest_unix.go:48: open /tmp/docker-graphtest-728787528/overlay: no such file or directory
20:25:53 --- FAIL: TestOverlayCreateBase (0.00s)
20:25:53 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-999998215/overlay: no such file or directory
20:25:53 	graphtest_unix.go:48: open /tmp/docker-graphtest-999998215/overlay: no such file or directory
20:25:53 --- FAIL: TestOverlayCreateSnap (0.00s)
20:25:53 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-711728314/overlay: no such file or directory
20:25:53 	graphtest_unix.go:48: open /tmp/docker-graphtest-711728314/overlay: no such file or directory
20:25:53 --- FAIL: TestOverlay50LayerRead (0.00s)
20:25:53 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-001075409/overlay: no such file or directory
20:25:53 	graphtest_unix.go:48: open /tmp/docker-graphtest-001075409/overlay: no such file or directory
20:25:53 FAIL
20:25:53 coverage: 10.3% of statements
20:25:53 FAIL	github.com/docker/docker/daemon/graphdriver/overlay	0.004s
20:25:53 --- FAIL: TestOverlaySetup (0.00s)
20:25:53 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-209180311/overlay2: no such file or directory
20:25:53 	graphtest_unix.go:48: open /tmp/docker-graphtest-209180311/overlay2: no such file or directory
20:25:53 --- FAIL: TestOverlayCreateEmpty (0.00s)
20:25:53 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-210387210/overlay2: no such file or directory
20:25:53 	graphtest_unix.go:48: open /tmp/docker-graphtest-210387210/overlay2: no such file or directory
20:25:53 --- FAIL: TestOverlayCreateBase (0.00s)
20:25:53 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-165683425/overlay2: no such file or directory
20:25:53 	graphtest_unix.go:48: open /tmp/docker-graphtest-165683425/overlay2: no such file or directory
20:25:53 --- FAIL: TestOverlayCreateSnap (0.00s)
20:25:53 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-280393420/overlay2: no such file or directory
20:25:53 	graphtest_unix.go:48: open /tmp/docker-graphtest-280393420/overlay2: no such file or directory
20:25:53 --- FAIL: TestOverlay128LayerRead (0.00s)
20:25:53 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-034209211/overlay2: no such file or directory
20:25:53 	graphtest_unix.go:48: open /tmp/docker-graphtest-034209211/overlay2: no such file or directory
20:25:53 --- FAIL: TestOverlayDiffApply10Files (0.01s)
20:25:53 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-956638912/overlay2: no such file or directory
20:25:53 	graphtest_unix.go:48: open /tmp/docker-graphtest-956638912/overlay2: no such file or directory
20:25:53 --- FAIL: TestOverlayChanges (0.00s)
20:25:53 	graphtest_unix.go:44: graphdriver: open /tmp/docker-graphtest-244229106/overlay2: no such file or directory
20:25:53 	graphtest_unix.go:48: open /tmp/docker-graphtest-244229106/overlay2: no such file or directory
20:25:53 FAIL
20:25:53 coverage: 13.5% of statements
```

AkihiroSuda commented:
It might be ok to leave the current snapshotters as-is, because we are going to integrate the containerd overlay driver in the next year: https://github.com/moby/buildkit/pull/160#discussion_r152439608

The containerd overlay driver already removed support for filesystems without d_type.

cc @dmcgowan @stevvooe

thaJeztah commented:
/cc @kolyshkin

kolyshkin commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -136,6 +136,9 @@ func Init(home string, options []string, uidMaps, gidMaps []idtools.IDMap) (grap
 
+	// Note: for simplicity, feature detection is done on /var/lib/docker
+	// this assumes that /var/lib/docker/overlay2 is on the same filesystem,
+	// and not a separate mount.
```
ditto

kolyshkin commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -171,6 +173,13 @@ func Init(home string, options []string, uidMaps, gidMaps []idtools.IDMap) (grap
 			}
 		}
 	}
+	supportsDType, err := fsutils.SupportsDType(home)
```
OK, you use `home` which is set to `/var/lib/docker/overlay2` by default and the directory might not exist yet. I guess a call to `filepath.Dir(home)` (same as in `graphdriver.GetFSMagic()`) is still needed either here or directly in `fsutils.SupportsDType()`.

kolyshkin commented on daemon/graphdriver/overlayutils/overlayutils.go:
```diff
@@ -13,6 +14,7 @@ func ErrDTypeNotSupported(driver, backingFs string) error {
 	}
-	msg += " Running without d_type support will no longer be supported in Docker 17.12."
-	return errors.New(msg)
+	msg += " Running without d_type support is not supported."
```
Perhaps `a filesystem not implementing d_type is not supported`?

thaJeztah replied:
Ah, yes, I can rephrase it a bit 👍

thaJeztah committed:
Perform fsmagic detection on driver's home-dir if it exists

The fsmagic check was always performed on "data-root" (`/var/lib/docker`),
not on the storage-driver's home directory (e.g. `/var/lib/docker/<somedriver>`).

This caused detection to be done on the wrong filesystem in situations
where `/var/lib/docker/<somedriver>` was a mount, and a different
filesystem than `/var/lib/docker` itself.

This patch checks if the storage-driver's home directory exists, and only
falls back to `/var/lib/docker` if it doesn't exist.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

kolyshkin reviewed (approved):
LGTM

thaJeztah commented:
There's one consequence of this PR; this check is not only done the first time the daemon starts (i.e., to find a working storage-driver), but also each time the daemon starts; even if `overlay2` (on a backing fs without `d_type`) was previously used.

So users that previously ran in such a setup, will now find out that their daemon refuses to start. Are we comfortable with that? (we _did_ warn it was no longer supported for some releases);

- do we need an `i-know-what-im-doing-get-me-outa-here` daemon option?
- do we need a different type of error ("unsupported-but-recoverable") that is printed as warning instead of fatal error when starting the daemon with an existing/already initialized `overlay` / `overlay2` graph-directory?

thaJeztah committed:
Allow existing setups to continue using d_type

Even though it's highly discouraged, there are existing
installs that are running overlay/overlay2 on filesystems
without d_type support.

This patch allows the daemon to start in such cases, instead of
refusing to start without an option to override.

For fresh installs, backing filesystems without d_type support
will still cause the overlay/overlay2 drivers to be marked as
"unsupported", and skipped during the automatic selection.

This feature is only to keep backward compatibility, but
will be removed at some point.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Added a commit that checks if the storage-driver home is already initialized, and if so, turns the error into a warning.

kolyshkin commented on daemon/graphdriver/driver.go:
```diff
@@ -277,6 +277,15 @@ func scanPriorDrivers(root string) map[string]bool {
+// IsInitialized checks if the driver's home-directory exists and is non-empty.
+func IsInitialized(driverHome string) bool {
+	if _, err := os.Stat(driverHome); os.IsNotExist(err) {
+		return false
```
It's not quite clear what to do with other errors from `os.Stat()`. On a normal system there should not be any, but if there is an error, it makes sense to at least report it to make debugging easier.

So, something like
```go
if err != nil {
   logrus.Warnf("graphdriver.IsInitialized: stat failed: %v", err)
}
```

kolyshkin commented on daemon/graphdriver/overlay/overlay.go:
```diff
@@ -148,7 +148,11 @@ func Init(home string, options []string, uidMaps, gidMaps []idtools.IDMap) (grap
+			return nil, overlayutils.ErrDTypeNotSupported("overlay", backingFs)
+		}
+		// allow running without d_type only for existing setups (#27443)
+		logrus.Warn(overlayutils.ErrDTypeNotSupported("overlay", backingFs))
```
Warnings are not logged unless daemon log-level is raised from the default, so maybe make it `logrus.Error()`?

thaJeztah replied:
Looks like they're logged;

```
root@9b94ad118dab:/go/src/github.com/docker/docker# dockerd --storage-driver=overlay2
INFO[2017-12-05T02:15:32.490617284Z] libcontainerd: started new docker-containerd process  pid=777
INFO[0000] starting containerd                           module=containerd revision=cc969fb42f427a68a8cc6870ef47f17304b83962 version=v1.0.0-rc.0-31-gcc969fb4
INFO[0000] setting subreaper...                          module=containerd
INFO[0000] changing OOM score to -500                    module=containerd
INFO[0000] loading plugin "io.containerd.content.v1.content"...  module=containerd type=io.containerd.content.v1
INFO[0000] loading plugin "io.containerd.snapshotter.v1.btrfs"...  module=containerd type=io.containerd.snapshotter.v1
WARN[0000] failed to load plugin io.containerd.snapshotter.v1.btrfs  error="path /var/lib/docker/containerd/daemon/io.containerd.snapshotter.v1.btrfs must be a btrfs filesystem to be used with the btrfs snapshotter" module=containerd
INFO[0000] loading plugin "io.containerd.snapshotter.v1.overlayfs"...  module=containerd type=io.containerd.snapshotter.v1
INFO[0000] loading plugin "io.containerd.metadata.v1.bolt"...  module=containerd type=io.containerd.metadata.v1
WARN[0000] could not use snapshotter btrfs in metadata plugin  error="path /var/lib/docker/containerd/daemon/io.containerd.snapshotter.v1.btrfs must be a btrfs filesystem to be used with the btrfs snapshotter" module="containerd/io.containerd.metadata.v1.bolt"
INFO[0000] loading plugin "io.containerd.differ.v1.walking"...  module=containerd type=io.containerd.differ.v1
INFO[0000] loading plugin "io.containerd.gc.v1.scheduler"...  module=containerd type=io.containerd.gc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.containers"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.content"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.diff"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.events"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.healthcheck"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.images"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.leases"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.namespaces"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.snapshots"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.monitor.v1.cgroups"...  module=containerd type=io.containerd.monitor.v1
INFO[0000] loading plugin "io.containerd.runtime.v1.linux"...  module=containerd type=io.containerd.runtime.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.tasks"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.version"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.introspection"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] serving...                                    address="/var/run/docker/containerd/docker-containerd-debug.sock" module="containerd/debug"
INFO[0000] serving...                                    address="/var/run/docker/containerd/docker-containerd.sock" module="containerd/grpc"
INFO[0000] containerd successfully booted in 0.014220s   module=containerd
WARN[2017-12-05T02:15:32.531397384Z] overlay: the backing extfs filesystem is formatted without d_type support, which leads to incorrect behavior. Backing filesystems without d_type support are not supported. 
INFO[2017-12-05T02:15:32.557281735Z] Graph migration to content-addressability took 0.00 seconds 
INFO[2017-12-05T02:15:32.560689230Z] Loading containers: start.                   
WARN[2017-12-05T02:15:32.561805372Z] Running modprobe bridge br_netfilter failed with message: , error: exec: "modprobe": executable file not found in $PATH 
WARN[2017-12-05T02:15:32.561933008Z] Running modprobe nf_nat failed with message: ``, error: exec: "modprobe": executable file not found in $PATH 
WARN[2017-12-05T02:15:32.562000374Z] Running modprobe xt_conntrack failed with message: ``, error: exec: "modprobe": executable file not found in $PATH 
INFO[2017-12-05T02:15:32.663232933Z] Default bridge (docker0) is assigned with an IP address 172.18.0.0/16. Daemon option --bip can be used to set a preferred IP address 
INFO[2017-12-05T02:15:32.706677300Z] Loading containers: done.                    
INFO[2017-12-05T02:15:32.767755745Z] Docker daemon                                 commit=04c1a99a0 graphdriver(s)=overlay2 version=dev
INFO[2017-12-05T02:15:32.769469688Z] Daemon has completed initialization          
INFO[2017-12-05T02:15:32.779885877Z] API listen on /var/run/docker.sock           
```

kolyshkin commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -184,7 +184,11 @@ func Init(home string, options []string, uidMaps, gidMaps []idtools.IDMap) (grap
+			return nil, overlayutils.ErrDTypeNotSupported("overlay2", backingFs)
+		}
+		// allow running without d_type only for existing setups (#27443)
+		logrus.Warn(overlayutils.ErrDTypeNotSupported("overlay2", backingFs))
```
ditto

thaJeztah commented:
Added a warning to the `IsInitialized()` function, but kept the other logs as Warning, because they seem to be logged; https://github.com/moby/moby/pull/35514#discussion_r154831082

kolyshkin reviewed (approved):
LGTM

vieux reviewed (approved):
LGTM
